src/shared/firewall-util-nft.c

```diff
@@ -1130,3 +1173,160 @@ int fw_nftables_add_local_dnat(
         /* table created anew; previous address already gone */
         return fw_nftables_add_local_dnat_internal(ctx->nfnl, add, af, protocol, local_port, remote, remote_port, NULL);
 }
+
+static const char *const nfproto_table[] = {
+        [NFPROTO_ARP]    = "arp",
+        [NFPROTO_BRIDGE] = "bridge",
+        [NFPROTO_INET]   = "inet",
+        [NFPROTO_IPV4]   = "ip",
+        [NFPROTO_IPV6]   = "ip6",
+        [NFPROTO_NETDEV] = "netdev",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(nfproto, int);
+
+void nft_set_context_free_many(NFTSetContexts *s) {
+        assert(s);
+
+        for (size_t i = 0; i < s->n_contexts; i++) {
+                free(s->contexts[i].table);
+                free(s->contexts[i].set);
+        }
+
+        s->n_contexts = 0;
+        s->contexts = mfree(s->contexts);
+}
+
+int nft_set_context_add(NFTSetContexts *s, int nfproto, const char *table, const char *set) {
+        _cleanup_free_ char *table_dup = NULL, *set_dup = NULL;
+        assert(s);
+        assert(IN_SET(nfproto, NFPROTO_ARP, NFPROTO_BRIDGE, NFPROTO_INET, NFPROTO_IPV4, NFPROTO_IPV6, NFPROTO_NETDEV));
+        assert(table);
+        assert(set);
+
+        table_dup = strdup(table);
+        if (!table_dup)
+                return -ENOMEM;
+
+        set_dup = strdup(set);
+        if (!set_dup)
+                return -ENOMEM;
+
+        NFTSetContext *c;
+        c = greedy_realloc((void **) &s->contexts, s->n_contexts + 1, sizeof(NFTSetContext));
```

Comment:
Why not to use `GREEDY_REALLOC()` ?

Reply:
It has a signature like `realloc()` but I'd need a function like `reallocarray()`, which is exactly what `greedy_realloc()` implements. What if I renamed `greedy_realloc()` to `greedy_reallocarray()`, introduced `GREEDY_REALLOCARRAY()` and used that?

Reply:
Sorry, I cannot follow. But it is not necessary to introduce another allocator.

Reply:
Sorry, I didn't check closely how GREEDY_REALLOC works and indeed it's great here.